modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/quantity_selector/QuantitySelector.js

```diff
@@ -12,175 +12,115 @@
-					console.log('stand alone component');
-				}
-			}
-		},
-		[props]
+function QuantitySelector({
+	componentId,
+	disabled,
+	forceDropdown,
+	name,
+	onUpdate,
+	quantity,
+	...optionSettings
+}) {
+	const [selectedQuantity, setSelectedQuantity] = useState(
+		optionSettings.minQuantity > quantity
+			? optionSettings.minQuantity
+			: quantity
 	);
 
-	const inputStyle = getInputStyle(props.inputSize);
+	const isDropdown =
+		optionSettings.allowedQuantities.length > 0 || forceDropdown;
+
+	/**
+	 * If source is <input /> and multipleQuantity > 1,
+	 * the newly set value will always be floored to the
+	 * closest lower multiple value.
+	 */
+	const onChange = ({target}) => {
+		if (target.value) {
+			const value = parseInt(target.value, 10);
 
-	useEffect(() => {
-		if (props.quantity) {
-			updateQuantity(props.quantity);
+			setSelectedQuantity(
+				value - (value % optionSettings.multipleQuantity)
+			);
 		}
 		else {
-			updateQuantity(selectedQuantity);
+			setSelectedQuantity(optionSettings.minQuantity || 1);
 		}
-	}, [selectedQuantity, props.quantity, updateQuantity]);
+	};
 
-	const content = (
-		<div
-			className="input-group input-group-sm quantity-selector simple"
-			data-testid="quantity-selector"
-			style={inputStyle}
-		>
-			{(props.prependedIcon || props.prependedText) && (
-				<div className="input-group-item input-group-item-shrink input-group-prepend">
-					<span className="input-group-text">
-						{props.prependedIcon ? (
-							<ClayIcon
-								spritemap={props.spritemap}
-								symbol={props.prependedIcon}
-							/>
-						) : (
-							props.prependedText
-						)}
-					</span>
-				</div>
-			)}
-			<div
-				className={classnames(
-					'input-group-item input-group-item-shrink',
-					(props.appendedIcon || props.appendedText) &&
-						'input-group-prepend'
-				)}
-			>
-				{props.style === 'datalist' && (
-					<Datalist
-						className="quantity-selector-input"
-						disabled={props.disabled}
-						id="quantitySelect"
-						size={props.size}
-						updateQuantity={updateQuantity}
-					>
-						{quantitiesList.map((item) => (
-							<option key={item} label={item} value={item} />
-						))}
-					</Datalist>
-				)}
+	const willUpdate = useCallback(() => {
+		clearTimeout(TIMEOUT_REF);
 
-				{props.style === 'select' && (
-					<ClaySelect
-						aria-label="Select Label"
-						classnames={classnames(
-							'quantity-selector-input',
-							props.size === 'small' && 'form-control-sm',
-							props.size === 'large' && 'form-control-lg'
-						)}
-						data-testid="select"
-						disabled={props.disabled}
-						id="quantitySelect"
-						onChange={(event) => {
-							updateQuantity(event.target.value);
-						}}
-					>
-						{quantitiesList.map((item) => (
-							<ClaySelect.Option
-								key={item}
-								label={item}
-								value={item}
-							/>
-						))}
-					</ClaySelect>
-				)}
-			</div>
+		TIMEOUT_REF = setTimeout(
```

Comment:
This shouldn't be all-caps; we usually use all-caps for _constants_.

Also, without really understanding this part of the code base, can there ever be more than one `QuantitySelector` on screen at a time? Seeing as they would be sharing the same `TIMEOUT_REF` reference... is that what you want?

Might be more "hygienic" to stick this in a `useRef` reference which is tied to the component instance.

```js
const timeout = useRef();

// ... and later...

timeout.current = setTimeout(...);
```

Reply:
Agreed for the useless caps. Yes, there will be multiple instances of quantity selectors in the page, and I didn't take the shared ref as an issue because, at the end of the day, it won't be possible to change > 1 Input/Select value at the same time. So by the time someone interacts with one of these component, I was just preferring to reuse the same "garbage" var not to pollute the component itself. What do you think? @wincent

Reply:
The reason "components" exist as a concept at all is for resuability. So, _even if they aren't actually reused in practice_, designing them for reusability makes them better.

As such, I would definitely make the component re-use-friendly by avoiding invisible shared global state. And while you say it's not possible to change more than one value at a time right now, making this change will make your code more resilient against possible changes in the future, if the constraints ever change (and they often do...)

tl;dr: It's not pollution, it's encapsulation.

Reply:
Oh btw, it's a short keypress debounce, if you're wondering.

Reply:
@wincent Alright! 👍